drivers/rtc/rtc_ds3231.c

```diff
@@ -0,0 +1,855 @@
+		return -ENODEV;
+	}
+
+	k_work_init(&data->work, rtc_ds3231_isw_h);
+
+	int err = gpio_pin_configure_dt(&(config->isw_gpios), GPIO_INPUT);
+
+	if (err != 0) {
+		LOG_ERR("Couldn't configure ISW GPIO pin.");
+		return err;
+	}
+	err = gpio_pin_interrupt_configure_dt(&(config->isw_gpios), GPIO_INT_EDGE_TO_ACTIVE);
+	if (err != 0) {
+		LOG_ERR("Couldn't configure ISW interrupt.");
+		return err;
+	}
+
+	gpio_init_callback(&data->isw_cb_data, rtc_ds3231_isw_isr, BIT((config->isw_gpios).pin));
+	err = gpio_add_callback((config->isw_gpios).port, &data->isw_cb_data);
+	if (err != 0) {
+		LOG_ERR("Couldn't add ISW interrupt callback.");
+		return err;
+	}
+
+	return 0;
+}
+#endif /* defined(CONFIG_RTC_UPDATE) || defined(CONFIG_RTC_ALARM) */
+
+static DEVICE_API(rtc, driver_api) = {
+	.set_time = rtc_ds3231_set_time,
+	.get_time = rtc_ds3231_get_time,
+
+#ifdef CONFIG_RTC_ALARM
+	.alarm_get_supported_fields = rtc_ds3231_alarm_get_supported_fields,
+	.alarm_set_time = rtc_ds3231_alarm_set_time,
+	.alarm_get_time = rtc_ds3231_alarm_get_time,
+	.alarm_is_pending = rtc_ds3231_alarm_is_pending,
+	.alarm_set_callback = rtc_ds3231_alarm_set_callback,
+#endif /* CONFIG_RTC_ALARM */
+
+#ifdef CONFIG_RTC_UPDATE
+	.update_set_callback = rtc_ds3231_update_set_callback,
+#endif /* CONFIG_RTC_UPDATE */
+
+#ifdef CONFIG_RTC_CALIBRATION
+/*.set_calibration = set_calibration,
+ * .get_calibration = get_calibration,
+ */
+#endif /* CONFIG_RTC_CALIBRATION */
+};
+
+static int rtc_ds3231_init_settings(const struct device *dev, const struct rtc_ds3231_conf *config)
+{
+	struct rtc_ds3231_settings conf = {
+		.osc = true,
+#ifdef CONFIG_RTC_UPDATE
+		.intctrl_or_sqw = false,
+		.freq_sqw = FREQ_1000,
+#else
+		.intctrl_or_sqw = true,
+#endif
+		.freq_32khz = config->freq_32k_gpios.port,
+	};
+	uint8_t mask = 255 & ~DS3231_BITS_STS_ALARM_1 & ~DS3231_BITS_STS_ALARM_2;
+	int err = rtc_ds3231_modify_settings(dev, &conf, mask);
+
+	if (err != 0) {
+		return err;
+	}
+	return 0;
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int rtc_ds3231_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	int err = 0;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND: {
+		struct rtc_ds3231_settings conf = {.osc = true,
+						   .intctrl_or_sqw = false,
+						   .freq_sqw = FREQ_1000,
+						   .freq_32khz = false};
+		uint8_t mask = 255 & ~DS3231_BITS_STS_ALARM_1 & ~DS3231_BITS_STS_ALARM_2;
+
+		err = rtc_ds3231_modify_settings(dev, &conf, mask);
+		if (err != 0) {
+			return err;
+		}
+		break;
+	}
+	case PM_DEVICE_ACTION_RESUME: {
+		/* TODO: trigger a temp CONV */
+		const struct rtc_ds3231_conf *config = dev->config;
+
+		err = rtc_ds3231_init_settings(dev, config);
+		if (err != 0) {
+			return err;
+		}
+		break;
+	}
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+#endif /* CONFIG_PM_DEVICE */
+
+static int rtc_ds3231_init(const struct device *dev)
+{
+	int err = 0;
+
+	const struct rtc_ds3231_conf *config = dev->config;
+	struct rtc_ds3231_data *data = dev->data;
+
```

Comment:
The status of the mfd device should still be checked, using the device_is_ready() function
```
if (!device_is_ready(config->mfd)) {
        return -ENODEV;
}
```

Reply:
same for the sensor driver